### ISSUE TITLE
chunk_limit_size not honored

### ISSUE THREAD

**Issue by uncycler** [uken/fluent-plugin-elasticsearch#535]
#### Problem

I don't know if it's a plugin or fluentd issue, but fluent-plugin-elasticsearch 3.0.1 included in td-agent 3.3.0 don't seem to honor "chunk_limit_size" buffer setting. After upgrading, fluentd stopped flushing buffer after some time because of this error:

2019-01-29 15:53:40 +0100 [warn]: #0 [elasticsearch] failed to flush the buffer. retry_time=1 next_retry_seconds=2019-01-29 15:54:10 +0100 chunk="580997a747fa17474251b90ace8f29e8" error_class=Fluent::Plugin::ElasticsearchOutput::RecoverableRequestFailure error="could not push logs to Elasticsearch cluster ({:host=>\"ece\", :port=>9200, :scheme=>\"http\", :user=>\"fluent\", :password=>\"obfuscated\"}): [413] Request Content-Length 528883194 exceeds the configured limit of 209715200"

Despite the fact that chunk_limit_size is set to 32MB

BTW, we use elasticsearch enterprise and the proxy is hardcoded to 209715200.

#### Steps to replicate

Upgrade td-agent to 3.3.0 and send lots of log

#### Expected Behavior or What you need to ask

Flush log at 32MB max

#### Using Fluentd and ES plugin versions

* OS version: CentOS 7.6
* VM
* td-agent 3.0.3
* ES plugin 3.0.1
* ECE 2.0, ES 6.5.1

**Comment by cosmo0920** [comment]
Could you try to add `total_limit_size 32MB` in your configuration?
`chunk_limit_size` sets up each of chunk size, but total chunk size is not limited by this parameter.

**Comment by uncycler** [comment]
`total_limit_size 32MB` would limit by whole buffer instance which is not wanted. My current buffer config looks like this:
```
<buffer>
  @type memory
  flush_thread_count 4
  flush_mode interval
  flush_interval 2s
  overflow_action drop_oldest_chunk
  retry_forever true
  retry_type periodic
  retry_wait 30s
  retry_randomize true
  chunk_limit_size 32MB
  compress gzip
  total_limit_size 12288MB
</buffer>
```

Would it be possible that a 32MB compressed chunk would create a 500MB json? But It never happened with older versions.

**Comment by cosmo0920** [comment]
Hmm..., could you try this patch?

```diff
diff --git a/lib/fluent/plugin/out_elasticsearch.rb b/lib/fluent/plugin/out_elasticsearch.rb
index 62b1986..e8fb677 100644
--- a/lib/fluent/plugin/out_elasticsearch.rb
+++ b/lib/fluent/plugin/out_elasticsearch.rb
@@ -51,6 +51,7 @@ module Fluent::Plugin
     DEFAULT_TYPE_NAME_ES_7x = "_doc".freeze
     DEFAULT_TYPE_NAME = "fluentd".freeze
     DEFAULT_RELOAD_AFTER = -1
+    TARGET_BULK_BYTES = 20 * 1024 * 1024
 
     config_param :host, :string,  :default => 'localhost'
     config_param :port, :integer, :default => 9200
@@ -526,6 +527,14 @@ EOC
 
           if append_record_to_messages(@write_operation, meta, header, record, bulk_message[info])
             bulk_message_count[info] += 1;
+            if bulk_message[info].size > TARGET_BULK_BYTES
+              bulk_message.each do |info, msgs|
+                send_bulk(msgs, tag, chunk, bulk_message_count[info], extracted_values, info.index) unless msgs.empty?
+                msgs.clear
+                # Clear bulk_message_count for this info.
+                bulk_message_count[info] = 0;
+              end
+            end
           else
             if @emit_error_for_missing_id
               raise MissingIdFieldError, "Missing '_id' field. Write operation is #{@write_operation}"
```

**Comment by cosmo0920** [comment]
> Would it be possible that a 32MB compressed chunk would create a 500MB json? But It never happened with older versions.

gzip compression reduces around 80% in most cases.
It is curious behavior.... :cold_sweat:

**Comment by cosmo0920** [comment]
Could you show use entire stack trace?

**Comment by uncycler** [comment]
it's all I got. The error is followed by the record entry.

**Comment by cosmo0920** [comment]
Umm....

How about the following patch? (`next` is added after `bulk_message_count[info] = 0`)

```diff
diff --git a/lib/fluent/plugin/out_elasticsearch.rb b/lib/fluent/plugin/out_elasticsearch.rb
index 62b1986..618766f 100644
--- a/lib/fluent/plugin/out_elasticsearch.rb
+++ b/lib/fluent/plugin/out_elasticsearch.rb
@@ -51,6 +51,7 @@ module Fluent::Plugin
     DEFAULT_TYPE_NAME_ES_7x = "_doc".freeze
     DEFAULT_TYPE_NAME = "fluentd".freeze
     DEFAULT_RELOAD_AFTER = -1
+    TARGET_BULK_BYTES = 20 * 1024 * 1024
 
     config_param :host, :string,  :default => 'localhost'
     config_param :port, :integer, :default => 9200
@@ -526,6 +527,15 @@ EOC
 
           if append_record_to_messages(@write_operation, meta, header, record, bulk_message[info])
             bulk_message_count[info] += 1;
+            if bulk_message.size > TARGET_BULK_BYTES
+              bulk_message.each do |info, msgs|
+                send_bulk(msgs, tag, chunk, bulk_message_count[info], extracted_values, info.index) unless msgs.empty?
+                msgs.clear
+                # Clear bulk_message_count for this info.
+                bulk_message_count[info] = 0;
+                next
+              end
+            end
           else
             if @emit_error_for_missing_id
               raise MissingIdFieldError, "Missing '_id' field. Write operation is #{@write_operation}"
```

**Comment by uncycler** [comment]
I've been running the patch on one of my node since friday and no error so far.

**Comment by cosmo0920** [comment]
Wow! Thank you for confiming my patch.
I'll send a PR to fix this issue. :smile:

**Comment by cosmo0920** [comment]
I've sent a PR to fix this issue.
And I modified my patch for minor fix which limits bulk_message contents length.
(Previous version just estimates bulk_message hash keys....)

**Comment by cosmo0920** [comment]
I've released a fix to this issue as v3.2.0.

**Comment by Nuru** [comment]
@cosmo0920 #539 is nice, but AWS limits chunk size to 10 MB, not 20 MB, so can we make this limit configurable?
